curated_transformers/util/string.py

```diff
@@ -0,0 +1,140 @@
+import re
+from abc import ABC, abstractmethod
+from typing import Optional, Tuple
+
+
+class StringTransform(ABC):
+    """
+    Base class for reversible string transformations.
+    """
+
+    def __init__(self, reversible: bool = True):
+        super().__init__()
+        self._reversible = reversible
+
+    @abstractmethod
+    def _apply(self, string: str) -> str:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _revert(self, string: str) -> str:
+        raise NotImplementedError
+
+    def apply(self, string: str) -> str:
+        """
+        Applies the transformation to the given string.
+
+        :param string:
+            String to transform.
+        :returns:
+            Transformed string.
+        """
+        return self._apply(string)
+
+    def revert(self, string: str) -> str:
+        """
+        Reverts the previously applied transformation of the given string.
+
+        :param string:
+            Previously transformed string.
+        :returns:
+            Reverted string.
+        """
+        if self._reversible:
+            return self._revert(string)
+        else:
+            return string
+
+
+class StrSub(StringTransform):
+    """
+    Substitute a substring with another string.
+    """
+
+    def __init__(self, forward: Tuple[str, str], backward: Optional[Tuple[str, str]]):
+        """
+        Construct a reversible substitution.
+
+        :param forward:
+            Tuple where the first string is a RegEx pattern
+            and the second the replacement.
+
+            This operation is performed when the :meth:`.apply`
+            method is invoked.
+        :param backward:
+            Optional tuple where the first string is a RegEx pattern
+            and the second the replacement.
+
+            This operation is performed when the :meth:`.revert`
+            method is invoked. If ``None``, it is a no-op.
+        """
+        super().__init__(backward is not None)
+
+        self.forward = forward
+        self.backward = backward
+
+    def _apply(self, string: str) -> str:
+        return re.sub(self.forward[0], self.forward[1], string)
+
+    def _revert(self, string: str) -> str:
+        assert self.backward is not None
+        return re.sub(self.backward[0], self.backward[1], string)
+
+
+class StrSubInv(StrSub):
```

Comment:
I'd say something like:

```suggestion
class StringSubRevesible(StrSub):
```

`Inv` is not really clear.

Reply:
Renamed to `StringSubInvertible` (since using reversible might be confusing since all transforms are allowed to be reversible).